crud/readview.lua

```diff
@@ -0,0 +1,322 @@
+                                                opts.sharding_key_hash,
+                                                opts.skip_sharding_hash_check)
+
+    if err ~= nil then
+        return nil, err
+    end
+
+    local filter_func, err = select_filters.gen_func(space, conditions, {
+        tarantool_iter = opts.tarantool_iter,
+        scan_condition_num = opts.scan_condition_num,
+    })
+    if err ~= nil then
+        return cursor, ReadviewError:new("Failed to generate tuples filter: %s", err)
+    end
+
+    -- execute select
+    local resp, err = select_executor.execute(space, index_format, filter_func, {
+        scan_value = opts.scan_value,
+        after_tuple = opts.after_tuple,
+        tarantool_iter = opts.tarantool_iter,
+        limit = opts.limit,
+        yield_every = opts.yield_every,
+        readview = true,
+        readview_index = index_readview,
+    })
+    if err ~= nil then
+        return cursor, ReadviewError:new("Failed to execute select: %s", err)
+    end
+
+    if resp.tuples_fetched < opts.limit or opts.limit == 0 then
+        cursor.is_end = true
+    else
+        local last_tuple = resp.tuples[#resp.tuples]
+        cursor.after_tuple = last_tuple
+    end
+
+    cursor.stats = {
+        tuples_lookup = resp.tuples_lookup,
+        tuples_fetched = resp.tuples_fetched,
+    }
+
+    -- getting tuples with user defined fields (if `fields` option is specified)
+    -- and fields that are needed for comparison on router (primary key + scan key)
+    local filtered_tuples = schema.filter_tuples_fields(resp.tuples, opts.field_names)
+
+    local result = {cursor, filtered_tuples}
+
+    local select_module_compat_info = stash.get(stash.name.select_module_compat_info)
+    if not select_module_compat_info.has_merger then
+        if opts.fetch_latest_metadata then
+            result[3] = cursor.storage_info.replica_schema_version
+        end
+    end
+
+    return unpack(result)
+end
+
+local Readview_obj = {}
+Readview_obj.__index = Readview_obj
+
+local select_call = stats.wrap(select.call, stats.op.SELECT)
+
+function Readview_obj:select(space_name, user_conditions, opts)
+    opts = opts or {}
+    opts.readview = true
+    opts.readview_uuid = self._uuid
+
+    return select_call(space_name, user_conditions, opts)
+end
+
+local pairs_call = stats.wrap(select.pairs, stats.op.SELECT, {pairs = true})
+function Readview_obj:pairs(space_name, user_conditions, opts)
+    opts = opts or {}
+    opts.readview = true
+    opts.readview_uuid = self._uuid
+
+    return pairs_call(space_name, user_conditions, opts)
+end
+
+function readview.init()
+    _G._crud.readview_open_on_storage = readview_open_on_storage
+    _G._crud.readview_close_on_storage = readview_close_on_storage
+    _G._crud.select_readview_on_storage = select_readview_on_storage
+ end
+
+function Readview_obj:close(opts)
+    checks('table', {
+        timeout = '?number',
+    })
+    opts = opts or {}
+    if self._opened == false then
```

Comment:
Let's also check `opened == true` on select and pairs and return error if closed. Add a small test case for it too.